src/main/scala/eu/proteus/job/operations/sax/SAXOperation.scala

```diff
@@ -0,0 +1,254 @@
+              case s2d: SensorMeasurement2D => s2d.x
+            }
+
+            (Math.min(acc._1, currentX), Math.max(acc._2, currentX))
+          }
+        }
+        out.collect((maxMin._1, maxMin._2, key))
+      }
+    }
+
+    val keyedStream = filteredStream
+      .keyBy(_.coilId)
+      .countWindow(this.numberOfPointsInPrediction())
+      .apply(reduceXFunction)
+
+    keyedStream.name("(MinX, MaxX, CoilId)")
+    keyedStream
+  }
+
+  /**
+   * Join the prediction results with the X window information.
+   * @param saxResults The SAX results.
+   * @param minMaxValues The minimum and maximum X values.
+   * @return A [[SAXResult]].
+   */
+  def joinResult(
+    saxResults: DataStream[SAXPrediction],
+    minMaxValues: DataStream[(Double, Double, Int)]) : DataStream[SAXResult] = {
+
+    val processFunction = new LinkOutputsFunction(this.targetVariable)
+
+    saxResults.keyBy(_.key).connect(minMaxValues.keyBy(_._3)).flatMap(processFunction)
+
+  }
+}
+
+
+class LinkOutputsFunction(targetVariable: String)
+  extends RichCoFlatMapFunction[SAXPrediction, (Double, Double, Int), SAXResult] {
+
+  @transient
+  private var coordsMap: MapState[Int, mutable.Queue[(Double, Double)]] = _
+
+  @transient
+  private var orphanPredictions: MapState[Int, mutable.Queue[SAXPrediction]] = _
+
+  override def open(parameters: Configuration): Unit = {
+    super.open(parameters)
+
+    val cfg = getRuntimeContext.getExecutionConfig
+    val keyType = createTypeInformation[Int]
+    val valueType1 = createTypeInformation[mutable.Queue[(Double, Double)]]
+    val valueType2 = createTypeInformation[mutable.Queue[SAXPrediction]]
+
+    val coordsDescriptor = new MapStateDescriptor[Int, mutable.Queue[(Double, Double)]](
+      "coords",
+      keyType.createSerializer(cfg),
+      valueType1.createSerializer(cfg)
+    )
+    val predictionsDescriptor = new MapStateDescriptor[Int, mutable.Queue[SAXPrediction]](
+      "sax-predictions",
+      keyType.createSerializer(cfg),
+      valueType2.createSerializer(cfg)
+    )
+    this.coordsMap = getRuntimeContext.getMapState(coordsDescriptor)
+    this.orphanPredictions = getRuntimeContext.getMapState(predictionsDescriptor)
+  }
+
+  /**
+   * Join a prediction with the associated coordinates.
+   * @param prediction The prediction information.
+   * @param coords The coordinates.
+   * @param out The output collector.
+   */
+  private def join(prediction: SAXPrediction, coords: (Double, Double), out: Collector[SAXResult]) : Unit = {
+    out.collect(new SAXResult(
+      prediction.key, this.targetVariable, coords._1, coords._2, prediction.classId, prediction.similarity))
+  }
+
+  override def flatMap2(in2: (Double, Double, Int), collector: Collector[SAXResult]) = {
+    val key = in2._3
+    if(this.orphanPredictions.contains(key) && this.orphanPredictions.get(key).nonEmpty){
+      val prediction = this.orphanPredictions.get(key).dequeue()
+      this.join(prediction, (in2._1, in2._2) , collector)
```

Comment:
Due to how rocksdb works, after we perform a dequeue, we need to update the state for the given key. I will provide a better solution after RP1. This should fix the bug reported by @0xNacho.